### ISSUE TITLE
InstantiationError for Abstract Nodes [DATAGRAPH-1470]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-data-neo4j#2031]
**[Joshua Mims](https://jira.spring.io/secure/ViewProfile.jspa?name=JIRAUSER50805)** opened **[DATAGRAPH-1470](https://jira.spring.io/browse/DATAGRAPH-1470?redirect=false)** and commented

Here is the basic setup:

 

```java
@Data
@Node
public class Person {  
    
    @Property 
    private String name;
    
    @Relationship(type = "HAS_PET")
    private List<Pet> pets;
}

@Data
@Node
public abstract class Pet {
    
    @Property 
    private String name;
}

@Data
@Node(labels = { "Dog", "Pet" })
public class Dog extends Pet {  

    @Property
    private String breed;
}

@Repository
public interface PetRepository extends Neo4jRepository<Pet, Long> {  

    @Query("MATCH (pet:Pet) WHERE pet.name = $name RETURN pet")
    findByName(String name);

    @Query("MATCH (pet:Pet)<-[has_pet:HAS_PET]-(person:Person) WHERE pet.name = $name RETURN pet, has_pet, person")
    findByNameWithOwner(String name);
}
```

Here we have some polymorphism.  `PetRepository::findByName` will return correctly, and the `Pet` node return should be an instance of `Dog`.  However, `PetRepository::findByNameWithOwner` will throw an InstantiationError for `Pet` (because `Pet` is abstract and cannot be instantiated).  After doing much digging into the framework, I believe this is because, in this instance, we are returning more than one object aside from the root ('Pet'), and in this case, `DefaultNeo4jEntityConverter::mergeRootNodeWithRecord` is called.  After this, the previous `MapAccessor` information is lost (most importantly, the `labels` property).  When this is passed on as `queryRoot` to the `map` method to later instantiate the object, it fails since there is no concrete type information left to create an instance from, and what's left is the abstract type

---

**Affects:** 6.0.2 (2020.0.2)

**Comment by meistermeier** [comment]
This is fixed by #2101
For completeness: The `@Id` definitions are missing and the relationship in the `Pet` if the custom query is correct.
Additionally if it should match the owner's name, the second custom query needs to be `s/pet.name/person.name`.

**Comment by myfrontpocket** [comment]
Updated to the latest version that includes these changes, but our repository call returns `null`, instead of blowing up with the instantiation error.
